dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/ResourceLink.java

```diff
@@ -95,50 +97,100 @@ public String toString() {
+            Host host = getHost((String)request.getAttribute(HOST_REQUEST_ATTRIBUTE) , user);
+            if(null == host){
+               host = getHost(contentlet.getHost(), user);
+            }
 
-                if(request.isSecure()){
-                    resourceLink.append(HTTPS_PREFIX);
-                }else{
-                    resourceLink.append(HTTP_PREFIX);
-                }
-                resourceLink.append(host.getHostname());
-                if(request.getServerPort() != 80 && request.getServerPort() != 443){
-                    resourceLink.append(StringPool.COLON).append(request.getServerPort());
-                }
-                resourceLinkUri.append(identifier.getParentPath()).append(contentlet.getStringProperty(FileAssetAPI.FILE_NAME_FIELD));
-                resourceLink.append(UtilMethods.encodeURIComponent(resourceLinkUri.toString()));
-                resourceLinkUri.append(LANG_ID_PARAM).append(contentlet.getLanguageId());
-                resourceLink.append(LANG_ID_PARAM).append(contentlet.getLanguageId());
+            if(request.isSecure()){
+                resourceLink.append(HTTPS_PREFIX);
+            }else{
+                resourceLink.append(HTTP_PREFIX);
+            }
+            resourceLink.append(host.getHostname());
+            if(request.getServerPort() != 80 && request.getServerPort() != 443){
+                resourceLink.append(StringPool.COLON).append(request.getServerPort());
+            }
+            resourceLinkUri.append(identifier.getParentPath()).append(binary.getName());
+            resourceLink.append(UtilMethods.encodeURIComponent(resourceLinkUri.toString()));
+            resourceLinkUri.append(LANG_ID_PARAM).append(contentlet.getLanguageId());
+            resourceLink.append(LANG_ID_PARAM).append(contentlet.getLanguageId());
 
-                final FileAsset fileAsset = getFileAsset(contentlet);
-                final String mimeType = fileAsset.getMimeType();
-                final String fileAssetName = fileAsset.getFileName();
+            final String mimeType = MimeTypeUtils.getMimeType(binary);
+            final String fileAssetName = binary.getName();
 
-                return new ResourceLink(resourceLink.toString(), resourceLinkUri.toString(), mimeType, fileAsset, isEditableAsText(mimeType, fileAssetName), downloadRestricted);
-            }
+            return new ResourceLink(resourceLink.toString(), resourceLinkUri.toString(), mimeType, contentlet,velocityVarName, isEditableAsText(mimeType, fileAssetName), downloadRestricted);
 
-            return new ResourceLink(StringPool.BLANK, StringPool.BLANK, StringPool.BLANK, null, false, true);
+            
         }
+        
+        public final ResourceLink buildFileAsset(final HttpServletRequest request, final User user, final Contentlet contentlet) throws DotDataException, DotSecurityException {
+
+            File binary = Try.of(()->contentlet.getBinary(FileAssetAPI.BINARY_FIELD)).getOrNull();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Local variable 'binary' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5337824)